### ISSUE TITLE
please provide a working example of automatic token refresh

### ISSUE THREAD

**Issue by bionicles** [googleapis/google-auth-library-nodejs#836]
**Is your feature request related to a problem? Please describe.**
I want to make a react app talk to an Apollo GraphQL server but I'm frustrated with the google oauth2 setup process because there's no example ... it's unclear from a big picture perspective

 the user clicks "Sign In" we get a code. We send the code to the server. Then what? 

if the offline flow only returns a code, how do you get the user's email and name and id without making a request to your server? 

Most of the examples use Express, with redirects, but GraphQL uses post, they're maintained by one random guy, they don't explain how to refresh the token / access user id on subsequent page views.

**Describe the solution you'd like**
How do you make sure the user doesn't have to sign in over and over and over? 

**What happens on subsequent signins?** 

Does the user have to re-consent or do we have to send a different button? 

Do you need to check if the user already consented and render different button settings? 

How do you get the id_token if you're doing offline flow?

Should I just give up on google and switch to Github or Facebook login?

Is there a CLIENT AND SERVER full-stack example of how to set this up that is actively maintained by google? 

Also, would it be possible to condense all the boilerplate?

I just want an official well maintained working example / developer guide for how to set up a react app and apollo graphql server so the user signs in ONE time and they don't have to re-sign-in hourly. Or, if offline flow is gonna be a huge mess, you guys should drop it, stop supporting it, and just simplify your docs so 50,000 devs don't waste multiple days trying and failing and trying and failing to set up single-sign-on with their react apps

Specifically, I want to:
1. import a google sign in button react component (right now, we have to use one guy's react component instead of officially maintained google code ... ) on my React app
2. when the user clicks it and consents to sign up for the app, I get the user's code to send to my graphql server, along with an id, name, and email.
3. With the id_token and access code, I want to fire a graphql mutation to save the refresh code, in the database with the id ? (or access ?) token as authorization header. 
4. clarify what happens the N+1 infinity times the user revisits the site. 

And I want it all to be published so other people don't have the same headaches I had

**Describe alternatives you've considered**
I tried cognito but it's super not-secure. 
Auth0 is a middleman, 
I'd use facebook or github and I probably should, because I'm frustrated with Google Sign In right now 
It's relatively doable to just make the user click on the button every hour, but that's also terrible UX

I just want to set up google sign in with react and graphql without days of annoying guesswork, 100s of lines of boilerplate, reading 50 random dude's blogs, not knowing which of 50 different google documentation pages to use. 
**Additional context**
https://www.apollographql.com/docs/apollo-server/security/authentication/
https://www.npmjs.com/package/react-google-login
https://github.com/anthonyjgrove/react-google-login/issues/175#
https://github.com/anthonyjgrove/react-google-login/pull/191
https://developers.google.com/identity/sign-in/web/server-side-flow (no nodejs server example, only java and python ... ?!)
https://developers.google.com/identity/sign-in/web/reference#googleauthgrantofflineaccessoptions
https://github.com/Shahzayb/mern-google-login/blob/master/controllers/auth.js

**Comment by bionicles** [comment]
https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/
https://developers.facebook.com/docs/facebook-login/
https://developer.twitter.com/en/docs/basics/authentication/overview/3-legged-oauth

**Comment by bionicles** [comment]
edited. sorry for long post. just want to get auth right. thanks in advance for any help

**Comment by bcoe** [comment]
@bionicles are you trying to make authenticated requests to Google APIs, or are you just trying to use Google as an OAuth provider for your application?

For authentication, I think reading up on OIDC would be a good start:

https://developers.google.com/identity/protocols/OpenIDConnect

:point_up: my hunch is that that document is more suited towards your use-case; this library helps make authenticated requests against google APIs, e.g., drive, cloudbuild, etc.

**Comment by bionicles** [comment]
yes and I have read this and am familiar with it,, and at the top of that link it sends you here: https://developers.google.com/identity/sign-in/web/backend-auth which shows clearly this 'google-auth-library-nodejs'  git repo as the solution to verify the identity on the server-side. thanks for confirming i posted this **issue with google-sign-in in nodejs** in the correct place! 

**the current solution has the following issues:**

1. you have to add a script tag to your html and you cannot add and import an npm package into your react app to put the sign in button. your customers do not want to load something from a cdn, we want to `yarn add google-signin`

2. developers get sucked into google conditional callback hell just to load and initialize the auth instance on the client side and frequently crashes the app if you use the code from the site you just linked. This is spaghetti boilerplate with tons of opportunity to screw up:
```
  useEffect(() => {
    if (authInstance) {
      console.log("authInstance exists");
      return;
    }
    if (!window.gapi) {
      console.log("no window.gapi");
      return;
    }
    window.gapi.load("client:auth2", () =>
      window.gapi.client.init({ clientId, scope }).then(initResponse => {
        window.gapi.auth2.getAuthInstance().then(authInstance => {
          console.log("authInstance", authInstance);
          dispatch(["SET", { authInstance }]);
          console.log("auth2 object:", window.gapi.auth2);
          const googler = authInstance.currentUser.get();
          console.log("googler", googler);
          const basicProfile = googler.getBasicProfile();
          console.log("basicProfile", basicProfile);
        });
      })
    );
  });
```

3. **every user has to suffer a context-switch and click twice every hour to login again because the access token times out and id verification stops working**. there's no working example of how to refresh the google sign-in token on the google sign-in docs; if you follow the current guide, you get to write a bunch of callback-hell boilerplate just so users must re-sign with google over and over and over and over and over. 

Honestly, if google sign-in / auth is a joke, it's a bad one, because as a developer you think, gee, google cybersecurity is gonna be simple and secure, but then the docs suck you into callback hell and your users must click double-digit times daily just to stay logged in, so you have access to their "profile and email" oauth2 scope. AND you have to trust kilobytes of new code which slows your load times! What a massive waste of time, energy, and money!

**Comment by vkopio** [comment]
I made a somewhat relevant [Stack Overflow question](https://stackoverflow.com/questions/61474363/trying-to-fetch-a-new-access-token-with-a-refresh-token-using-google-auth-librar) as I am having similar problems with the `refresh_token`.

**Comment by renanlecaro** [comment]
I'm also frustrated by the lack of examples in how to handle the token refresh workflow. My app ends up re-prompting users all the time because the refresh fails very often. I know i'm the one that didn't code it right, but as this is my first implementation of an OAuth2 flow, i'd love more documentation on the token refresh process

**Comment by bcoe** [comment]
we will make an effort to add docs that demonstrate the refresh token flow 👍

**Comment by bennettrogers** [comment]
I agree with everything said here by @bionicles and @renanlecaro. I've spent far too many hours recently trying to implement the OAuth flow to access Google APIs on behalf of users. There are multiple competing "official" Google node.js SDKs (at [least this one](https://github.com/googleapis/google-auth-library-nodejs) and [this one](https://github.com/googleapis/google-api-nodejs-client/)), yet as @bionicles mentions, the only dev-friendly React components for Google Auth are a couple of repos occasionally maintained by solo devs (thank you to @anthonyjgrove and @asyarb for their [two](https://github.com/anthonyjgrove/react-google-login) [packages](https://github.com/asyarb/react-use-googlelogin).

This feels like yet another case of too many Google PM cooks in the kitchen, "landing" competing projects to get promoted, while nobody looks at the bigger picture.

I finally managed to get the complete flow working, and then find that when I go to verify the jwt id_token on subsequent requests, the verification throws an error since the id_token has expired. Neat.

I am left with 2 bad options: either stop verifying the id_token entirely, or hack the verification to detect "token used too late" errors and ignore them.

I guess I should include a question - which of these options is the recommended official approach?

This is not a problem with Auth0 - Auth0 just works. I feel like Google probably has the resources to do better, but chooses to prioritize other things.
